math/minuit2/src/NumericalDerivator.cxx

```diff
@@ -0,0 +1,336 @@
+                                                       unsigned int i_component, DerivatorElement previous)
+{
+   SetupDifferentiate(function, x, parameters);
+   return FastPartialDerivative(function, parameters, i_component, previous);
+}
+
+// leaves the parameter setup to the caller
+DerivatorElement NumericalDerivator::FastPartialDerivative(const ROOT::Math::IBaseFunctionMultiDim *function,
+                                                           const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                                                           unsigned int i_component, const DerivatorElement &previous)
+{
+   DerivatorElement deriv{previous};
+
+   double xtf = fVx[i_component];
+   double epspri = fPrecision.Eps2() + std::abs(deriv.derivative * fPrecision.Eps2());
+   double step_old = 0.;
+
+   for (unsigned int j = 0; j < fNCycles; ++j) {
+      double optstp = std::sqrt(fDfmin / (std::abs(deriv.second_derivative) + epspri));
+      double step = std::max(optstp, std::abs(0.1 * deriv.step_size));
+
+      if (parameters[i_component].IsBound()) {
+         if (step > 0.5)
+            step = 0.5;
+      }
+
+      double stpmax = 10. * std::abs(deriv.step_size);
+      if (step > stpmax)
+         step = stpmax;
+
+      double stpmin = std::max(fVrysml, 8. * std::abs(fPrecision.Eps2() * fVx[i_component]));
+      if (step < stpmin)
+         step = stpmin;
+      if (std::abs((step - step_old) / step) < fStepTolerance) {
+         break;
+      }
+      deriv.step_size = step;
+      step_old = step;
+      fVx[i_component] = xtf + step;
+      fVxExternal[i_component] = Int2ext(parameters[i_component], fVx[i_component]);
+      double fs1 = (*function)(fVxExternal.data());
+
+      fVx[i_component] = xtf - step;
+      fVxExternal[i_component] = Int2ext(parameters[i_component], fVx[i_component]);
+      double fs2 = (*function)(fVxExternal.data());
+
+      fVx[i_component] = xtf;
+      fVxExternal[i_component] = Int2ext(parameters[i_component], fVx[i_component]);
+
+      double fGrd_old = deriv.derivative;
+      deriv.derivative = 0.5 * (fs1 - fs2) / step;
+
+      deriv.second_derivative = (fs1 + fs2 - 2. * fVal) / step / step;
+
+      if (std::abs(fGrd_old - deriv.derivative) / (std::abs(deriv.derivative) + fDfmin / step) < fGradTolerance) {
+         break;
+      }
+   }
+   return deriv;
+}
+
+DerivatorElement NumericalDerivator::operator()(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                                                const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                                                unsigned int i_component, const DerivatorElement &previous)
+{
+   return PartialDerivative(function, x, parameters, i_component, previous);
+}
+
+std::vector<DerivatorElement>
+NumericalDerivator::Differentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *cx,
```

Comment:
```suggestion
NumericalDerivator::Differentiate(const ROOT::Math::IBaseFunctionMultiDim &function, const double *cx,
```
`function` can not be a `nullptr` right? In this case, the function parameters could be taken by reference to make the code clearer.

Reply:
Ah, I have to check to be sure, but I think you're right. This was probably again a historical thing. There's even an assert checking for nullptr in SetupDifferentiate, so seems highly unlikely that nullptrs still make it through... Ah yeah, I think I remember now that I used to also have a copy of the function (pointer) in the class itself, but then for the newer class I needed to pass it in dynamically, so I added a check somewhere for nullptr calls, I think, that would use the internal function pointer...

Ok, so I think it can indeed be turned into a reference. However, I'm not completely confident that this won't mess up anything with the complicated inheritance hierarchy and pointer casting magic in #8596... Do you think it's ok to postpone this change?

Reply:
Sure, we can do it later indeed!